### PR TITLE
Tech group mod2 final mary luna

### PR DESCRIPTION
**Pull request by jimenaluna84**

### CODE REVIEW

File: src/lunamary/ReadWriteData/file.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
+[
```
Comment (JurgenPadilla) [inline src/lunamary/ReadWriteData/file.json:1]: Please, move these kinds of files to a package named resources under you directory name.

File: src/lunamary/modelPerson/Student.java [hunk 2]
```diff
@@ -1,15 +1,22 @@
 package lunamary.modelPerson;
 
-import java.util.List;
+public class Student extends AbstractPerson {
+
 
-public class Student extends Person {
     private Parent parent;
```
Comment (JurgenPadilla) [inline src/lunamary/modelPerson/Student.java:6]: probably a student can have more than one parent

File: src/lunamary/modelPerson/Teacher.java [hunk 3]
```diff
@@ -1,11 +1,9 @@
 package lunamary.modelPerson;
 
-import java.util.List;
+public class Teacher extends AbstractPerson {
```
Comment (JurgenPadilla) [inline src/lunamary/modelPerson/Teacher.java:3]: You should add more attributes to this class.

File: src/lunamary/services/ClassroomService.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+import lunamary.modelPerson.Student;
+import lunamary.modelSchool.Classroom;
+
+public class ClassroomService {
+
+    public ClassroomService() {
+    }
+
+    public Classroom crateClassroom(String id, String name) {
```
Comment (JurgenPadilla) [inline src/lunamary/services/ClassroomService.java:11]: fix typo

File: src/lunamary/services/CommonService.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+import datastructures.circulardoublylinkedlist.MyCircularDoublyLinkedList;
+import lunamary.modelPerson.Parent;
+import lunamary.modelPerson.Student;
+import lunamary.modelPerson.Teacher;
+import lunamary.modelSchool.Classroom;
+import lunamary.modelSchool.School;
+import lunamary.modelSchool.Subject;
+
+public class CommonService {
```
Comment (JurgenPadilla) [inline src/lunamary/services/CommonService.java:11]: You need to rename this class to something more explanatory.

File: src/lunamary/services/CommonService.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+
+    public CommonService() {
+
+    }
+
+    public void print(String text) {
```
Comment (JurgenPadilla) [inline src/lunamary/services/CommonService.java:18]: Not sure why are you mixing these prints with service data but is not ok to do that.

File: src/lunamary/services/SchoolService.java [hunk 7]
```diff
@@ -1,4 +1,176 @@
...
+    private static School school;
+    private AbstractFactory factory;
+
+
+    public SchoolService() {
+
+
+    }
+
+    public School createSchool(String name, String address) {
```
Comment (JurgenPadilla) [inline src/lunamary/services/SchoolService.java:26]: This is not a proper implementation of Singleton design pattern, please check some documentation.
Implementing this correctly will simplify a lot of the logic

File: src/lunamary/services/SchoolService.java [hunk 8]
```diff
@@ -1,4 +1,176 @@
...
+        Director director = directorService.createDirector(name, lastname, ci);
+        school.setDirector(director);
+
+    }
+
+    public static void registerClassroom(String id, String name) {
+        ClassroomService classroomService = new ClassroomService();
+        Classroom classroom = classroomService.crateClassroom(id, name);
+        school.addClassroom(classroom);
+        System.out.println(classroom.getName().toString());
```
Comment (JurgenPadilla) [inline src/lunamary/services/SchoolService.java:42]: Please, do not mix Service logic with Presentation.

File: src/lunamary/ReadWriteData/AbstractFactory.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
+package lunamary.ReadWriteData;
+
+public class AbstractFactory {
```
Comment (JurgenPadilla) [inline src/lunamary/ReadWriteData/AbstractFactory.java:3]: You can't name your class Abstract and the class not being abstract actually.

File: src/lunamary/ReadWriteData/AbstractFactory.java [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+    public ReadWriteFile createFile(String path) {
```
Comment (JurgenPadilla) [inline src/lunamary/ReadWriteData/AbstractFactory.java:5]: Factory methods are usually static so there is no need to instantiate objects of these classes to make use of them

File: src/lunamary/ReadWriteData/ReadWriteCSV.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+            while (line != null) {
+                MyHashMap<String, String> entry = new MyHashMap<>();
+                for (int i = 0; i < headers.length; i++) {
+                    entry.put(headers[i], line[i]);
+                }
+                entries.add(entry);
+                line = csvReader.readNext();
+            }
+        } catch (Exception exception) {
+            System.out.println(exception);
```
Comment (JurgenPadilla) [inline src/lunamary/ReadWriteData/ReadWriteCSV.java:32]: Make the method throw this exception, as I said on other comments you are mixing the service logic with presentation.

File: src/lunamary/ReadWriteData/ReadWriteJSON.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        } catch (Exception exception) {
+            System.out.println(exception);
+        }
+
+        return records;
+    }
+
+    @Override
+    public boolean writeEntries(MyArrayList<MyHashMap<String, String>> lines) {
+        return false;
```
Comment (JurgenPadilla) [inline src/lunamary/ReadWriteData/ReadWriteJSON.java:45]: Same as above

File: src/lunamary/modelPerson/Device.java [hunk 13]
```diff
@@ -2,13 +2,15 @@
 
 public class Device {
 
-    private String type; // App, Movil, Email
-    private String identifier; // number or email
+    private String type;
+    private String identifier;
```
Comment (JurgenPadilla) [inline src/lunamary/modelPerson/Device.java:6]: Looks like you need identifiers on most of your entities. Probably is a good idea to create a base abstract class called BaseModel or something that has a unique identifier you can use java.util.UUID to achieve that

File: src/lunamary/services/CommonService.java [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+        System.out.print(text);
+    }
+
+
+    public static Classroom getClassroom(School school, String codeClassroom) {
```
Comment (JurgenPadilla) [inline src/lunamary/services/CommonService.java:23]: If you implement singleton correctly there is no need to pass school entities all over the place.

**Review by JurgenPadilla — CHANGES_REQUESTED** [review]

File: src/lunamary/ReadWriteData/file.json (hunk 1 above)
Comment (jimenaluna84) [reply src/lunamary/ReadWriteData/file.json:1]: Done!

**Pull request closed without merging**